### PR TITLE
feat(BRepQueries): Added queries to access the facet(s) of a block co…

### PR DESCRIPTION
**Pull request by MelchiorSchuh**
…rresponding to the polygon of a surface of a BRep.

### CODE REVIEW

File: src/geode/model/helpers/component_mesh_queries.cpp [hunk 1]
```diff
@@ -137,4 +217,77 @@ namespace geode
             "triangles and quadrangles." );
         return {};
     }
+
+    absl::InlinedVector< PolygonVertices, 2 >
+        block_vertices_from_surface_polygon( const BRep& model,
```
Comment (BotellaA) [inline src/geode/model/helpers/component_mesh_queries.cpp:222]: We need to discuss what this function does. Looks too much according to its name.

File: src/geode/model/helpers/component_mesh_queries.cpp [hunk 2]
```diff
@@ -109,10 +112,87 @@ namespace
...
+        {
+            const auto block_vertex_indices = model.mesh_component_vertices(
+                polygon_unique_vertices[polygon_vertex], block.id() );
+            if( block_vertex_indices.size() == 1 )
+            {
+                block_facet_vertices.emplace_back( block_vertex_indices[0] );
+            }
+            else
+            {
+                return absl::nullopt;
```
Comment (panquez) [inline src/geode/model/helpers/component_mesh_queries.cpp:134]: So it does no return something if surface is internal?

File: src/geode/model/helpers/component_mesh_queries.cpp [hunk 3]
```diff
@@ -137,4 +217,77 @@ namespace geode
...
+                    facets_block_vertices.back() );
+                std::swap( facets_unique_vertices.front(),
+                    facets_unique_vertices.back() );
+            }
+            else
+            {
+                OPENGEODE_ASSERT( polygon_unique_vertices_cycle.is_opposite(
+                                      { facets_unique_vertices[1] } ),
+                    "[block_vertices_from_surface_polygon] The block facets "
+                    "found from the polygon vertices have the same normal." );
```
Comment (panquez) [inline src/geode/model/helpers/component_mesh_queries.cpp:266]:
```suggestion
                    "found from the polygon vertices have the same orientation." );
```

File: src/geode/model/helpers/component_mesh_queries.cpp (hunk 1 above)
Comment (panquez) [reply src/geode/model/helpers/component_mesh_queries.cpp:222]: and maybe the function is a bit too long

File: src/geode/model/helpers/component_mesh_queries.cpp (hunk 2 above)
Comment (MelchiorSchuh) [reply src/geode/model/helpers/component_mesh_queries.cpp:134]: If the surface is internal or if the a polygon vertex has several mesh_component_vertices associated, it returns a null_opt (e.g. it cannot find the vertices from this information only).

File: include/geode/model/helpers/component_mesh_queries.h [hunk 4]
```diff
@@ -26,20 +26,65 @@
...
 
 namespace geode
 {
+    FORWARD_DECLARATION_DIMENSION_CLASS( Surface );
+    ALIAS_3D( Surface );
     FORWARD_DECLARATION_DIMENSION_CLASS( Block );
     ALIAS_3D( Block );
     class BRep;
+
+    struct BlockPolyhedronsFacetVertices
```
Comment (BotellaA) [inline include/geode/model/helpers/component_mesh_queries.h:41]: We used to declare this not in the namespace for forward declarations.

File: src/geode/model/helpers/component_mesh_queries.cpp [hunk 5]
```diff
@@ -137,4 +162,177 @@ namespace geode
...
+            "is neither boundary nor internal to the given block in the given "
+            "model." );
+        const auto polygon_unique_vertices =
+            surface_polygon_unique_vertices( model, surface, polygon_id );
+        BlockPolyhedronsFacetVertices polygon_to_block_vertices;
+        if( auto block_vertices =
+                block_unique_facet_vertices_from_surface_polygon(
+                    model, block, polygon_unique_vertices ) )
+        {
+            polygon_to_block_vertices.oriented_polyhedron_vertices =
```
Comment (BotellaA) [inline src/geode/model/helpers/component_mesh_queries.cpp:297]: Why oriented by default ?

File: include/geode/model/helpers/component_mesh_queries.h (hunk 4 above)
Comment (panquez) [inline include/geode/model/helpers/component_mesh_queries.h:41]:
```suggestion
    struct BlockPolyhedraFacetVertices
```
?

File: src/geode/model/helpers/component_mesh_queries.cpp [hunk 6]
```diff
@@ -137,4 +162,234 @@ namespace geode
...
+            "is neither boundary nor internal to the given block in the given "
+            "model." );
+        const auto polygon_unique_vertices =
+            surface_polygon_unique_vertices( model, surface, polygon_id );
+        BlockPolyhedraFacetVertices p_to_block_vertices;
+        if( auto block_vertices =
+                block_unique_facet_vertices_from_surface_polygon(
+                    model, block, polygon_unique_vertices ) )
+        {
+            p_to_block_vertices.oriented_polyhedron_vertices =
```
Comment (BotellaA) [inline src/geode/model/helpers/component_mesh_queries.cpp:354]: Why still oriented by default? Why not have the block_mesh_polyhedra_from_surface_polygon return a struct telling you which facet is oriented or not?

Comment (BotellaA) [inline src/geode/model/helpers/component_mesh_queries.cpp:354]: Why still oriented by default? Why not have the block_mesh_polyhedra_from_surface_polygon return a struct telling you which facet is oriented or not?

File: include/geode/model/helpers/component_mesh_queries.h [hunk 7]
```diff
@@ -26,20 +26,68 @@
...
+                counter++;
+            }
+            if( opposite_polyhedron_vertices )
+            {
+                counter++;
+            }
+            return counter;
+        }
+
+        absl::optional< PolygonVertices > any_of() const
```
Comment (panquez) [inline include/geode/model/helpers/component_mesh_queries.h:60]: Not used anymore?

File: src/geode/model/helpers/component_mesh_queries.cpp [hunk 8]
```diff
@@ -137,4 +162,234 @@ namespace geode
...
+    }
+
+    /*!
+     * Returns the result of polygon_vertices_to_block_facets_vertices, but with
+     * the lists oriented so that the first PolygonVertices of each vector is
+     * linked to the facet with the same orientation as the given vertices.
+     */
+    std::tuple< absl::InlinedVector< geode::PolygonVertices, 2 >,
+        absl::InlinedVector< geode::PolygonVertices, 2 >,
+        bool >
```
Comment (panquez) [inline src/geode/model/helpers/component_mesh_queries.cpp:209]: maybe i am wrong, but why do not you use `BlockPolyhedraFacetVertices` as return type?

**Review by panquez — APPROVED** [review]
I have a bit of difficulty to understand all the cpp code, so my comments are maybe wrong

File: src/geode/model/helpers/component_mesh_queries.cpp (hunk 8 above)
Comment (MelchiorSchuh) [reply src/geode/model/helpers/component_mesh_queries.cpp:209]: Yes, they're not the same object, this function returns two lists: the PolygonVertices representing the vertices in the block of each facet, and the PolygonVertices rerpesenting the unique vertices associated (with a corresponding numbering and orientation in the two lists.

**Review by BotellaA — APPROVED** [review]

**Pull request merged**